### PR TITLE
Ndev 3173 create tests to verify events

### PR DESCRIPTION
**Pull request by andrei-neon**

### CODE REVIEW

File: integration/tests/basic/indexer/test_events.py [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+    count_events,
+)
+from utils.accounts import EthAccounts
+from utils.helpers import wait_condition
+from utils.models.result import NeonGetTransactionResult
+from utils.web3client import NeonChainWeb3Client
+
+
+@pytest.fixture(scope="class", params=["neon", "sol"])
+def client_and_price(web3_client, web3_client_sol, sol_price, neon_price, request, pytestconfig):
```
Comment (kristinaNikolaevaa) [inline integration/tests/basic/indexer/test_events.py]: it's needed for testing multi chain transactions
You can use only web3_client for this tests
and you don't need price at all

File: integration/tests/basic/indexer/test_events.py [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+
+@allure.feature("JSON-RPC validation")
+@allure.story("Verify events")
+@pytest.mark.usefixtures("accounts", "web3_client")
+class TestEvents:
+    web3_client: NeonChainWeb3Client
+    accounts: EthAccounts
+
+    @pytest.mark.mainnet
+    @pytest.mark.neon_only
```
Comment (kristinaNikolaevaa) [inline integration/tests/basic/indexer/test_events.py]: please set neon mark for the class and delete mainnet mark

Comment (kristinaNikolaevaa) [reply integration/tests/basic/indexer/test_events.py]: for all tests

File: integration/tests/basic/indexer/test_events.py [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+        validated_response = NeonGetTransactionResult(**response)
+
+        transaction_hash = resp["transactionHash"].hex()
+        assert_event_field(validated_response, NeonEventType.EnterCall, "transactionHash", transaction_hash)
+        assert_event_field(validated_response, NeonEventType.ExitStop, "transactionHash", transaction_hash)
+        assert_event_field(validated_response, NeonEventType.Return, "transactionHash", transaction_hash)
+
+        assert_event_by_type(validated_response, NeonEventType.EnterCall)
+        assert_event_by_type(validated_response, NeonEventType.ExitStop)
+        assert_event_by_type(validated_response, NeonEventType.Return)
```
Comment (kristinaNikolaevaa) [inline integration/tests/basic/indexer/test_events.py]: you have the same checks with the same transaction in previous tests. Maybe better to merge these tests

File: integration/tests/basic/indexer/test_events.py [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+        assert_event_by_type(validated_response, NeonEventType.ExitRevert)
+
+    @pytest.mark.mainnet
+    @pytest.mark.neon_only
+    def test_contract_interative_operation(
+        self, counter_contract, client_and_price, account_with_all_tokens, json_rpc_client, operator
+    ):
+        w3_client, *_ = client_and_price
+
+        sol_balance_before = operator.get_solana_balance()
```
Comment (kristinaNikolaevaa) [inline integration/tests/basic/indexer/test_events.py]: why do you need operator balance for this test?

File: integration/tests/basic/indexer/test_events_by_type.py [hunk 5]
```diff
@@ -0,0 +1,335 @@
...
+)
+from utils.accounts import EthAccounts
+from utils.models.result import NeonGetTransactionResult
+from utils.web3client import NeonChainWeb3Client
+
+
+@allure.feature("JSON-RPC validation")
+@allure.story("Verify events")
+@pytest.mark.usefixtures("accounts", "web3_client")
+class TestNeonEventType:
```
Comment (kristinaNikolaevaa) [inline integration/tests/basic/indexer/test_events_by_type.py]: I don't see the difference between test_events and tests_events_by_type
looks like they do the same but for different set of events

File: integration/tests/basic/indexer/test_events_by_type.py [hunk 6]
```diff
@@ -0,0 +1,335 @@
...
+        assert_events_order(validated_response)
+        assert_event_by_type(validated_response, NeonEventType.EnterCallCode)
+        assert_event_by_type(validated_response, NeonEventType.EnterCall)
+        assert_event_by_type(validated_response, NeonEventType.ExitReturn)
+        assert_event_by_type(validated_response, NeonEventType.ExitStop)
+        assert_event_by_type(validated_response, NeonEventType.Return)
+
+    @pytest.mark.mainnet
+    @pytest.mark.neon_only
+    def test_event_enter_static_call(self, json_rpc_client, tracer_caller_contract, tracer_callee_contract_address):
```
Comment (kristinaNikolaevaa) [inline integration/tests/basic/indexer/test_events_by_type.py]: let's move tracer_caller_contract, tracer_callee_contract_address from tracer folder and rename to events checker or something like that

File: integration/tests/basic/indexer/test_events_by_type.py [hunk 7]
```diff
@@ -0,0 +1,335 @@
...
+        assert_event_by_type(validated_response, NeonEventType.Return)
+
+    @pytest.mark.mainnet
+    @pytest.mark.neon_only
+    def test_event_enter_call(self, json_rpc_client, event_caller_contract):
+        sender_account = self.accounts[0]
+        tx = self.web3_client.make_raw_tx(sender_account)
+        instruction_tx = event_caller_contract.functions.nonArgs().build_transaction(tx)
+        resp = self.web3_client.send_transaction(sender_account, instruction_tx)
+        response = json_rpc_client.send_rpc(method="neon_getTransactionReceipt", params=[resp["transactionHash"].hex()])
```
Comment (kristinaNikolaevaa) [inline integration/tests/basic/indexer/test_events_by_type.py]: let's move it to json_rpc_client to the new function like get_neon_trx_receipt

File: integration/tests/basic/helpers/rpc_checks.py [hunk 8]
```diff
@@ -201,4 +206,172 @@ def assert_equal_fields(result, comparable_object, comparable_fields, keys_mappi
...
+):
+    events = neon_trx_receipt.get_all_events(ignore_events=ignored_events, is_removed=is_removed)
+    return Counter([event.neonEventType for event in events])
+
+
+def assert_event_by_type(neon_trx_receipt: NeonGetTransactionResult, event_type):
+    events = neon_trx_receipt.get_all_events(filter_by_type=event_type)
+
+    for event in events:
+        match event_type:
```
Comment (kristinaNikolaevaa) [inline integration/tests/basic/helpers/rpc_checks.py]: maybe better to delete match event_type and check all events what neon_trx_receipt has in loop ?

File: integration/tests/basic/indexer/conftest.py [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+        "common/NestedCallsChecker", "0.8.12", accounts[0], contract_name="B"
+    )
+    contract_c, _ = web3_client.deploy_and_get_contract(
+        "common/NestedCallsChecker", "0.8.12", accounts[0], contract_name="C"
+    )
+    yield contract_a, contract_b, contract_c
+
+
+@pytest.fixture(scope="function")
+def recursion_factory(accounts, web3_client):
```
Comment (kristinaNikolaevaa) [inline integration/tests/basic/indexer/conftest.py]: lets move all common contracts to common conftest files no to duplicate the fixtures

File: integration/tests/basic/indexer/test_events.py [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+        tx = w3_client.make_raw_tx(account_with_all_tokens.address)
+
+        instruction_tx = counter_contract.functions.moreInstruction(0, 1000).build_transaction(tx)
+        resp = w3_client.send_transaction(account_with_all_tokens, instruction_tx)
+        wait_condition(lambda: sol_balance_before > operator.get_solana_balance())
+        response = json_rpc_client.send_rpc(method="neon_getTransactionReceipt", params=[resp["transactionHash"].hex()])
+        validated_response = NeonGetTransactionResult(**response)
+        assert len(validated_response.result.solanaTransactions) > 1
+        assert_events_order(validated_response)
```
Comment (kristinaNikolaevaa) [inline integration/tests/basic/indexer/test_events.py:193]: Why don't you check event count here?

**Review by kristinaNikolaevaa — CHANGES_REQUESTED** [review]

File: integration/tests/basic/indexer/test_events.py (hunk 2 above)
Comment (andrei-neon) [reply integration/tests/basic/indexer/test_events.py]: Fixed ✅

File: integration/tests/basic/indexer/test_events.py (hunk 1 above)
Comment (andrei-neon) [reply integration/tests/basic/indexer/test_events.py]: Removed ✅

File: integration/tests/basic/indexer/test_events.py (hunk 3 above)
Comment (andrei-neon) [reply integration/tests/basic/indexer/test_events.py]: Done ✅

File: integration/tests/basic/indexer/test_events.py (hunk 4 above)
Comment (andrei-neon) [reply integration/tests/basic/indexer/test_events.py]: It was used for the `wait_condition()` function but looks like there is no need for it at all. Removed ✅

File: integration/tests/basic/indexer/test_events_by_type.py (hunk 5 above)
Comment (andrei-neon) [reply integration/tests/basic/indexer/test_events_by_type.py]: Refactored ✅

File: integration/tests/basic/indexer/test_events_by_type.py (hunk 6 above)
Comment (andrei-neon) [reply integration/tests/basic/indexer/test_events_by_type.py]: Fixed ✅

File: integration/tests/basic/indexer/test_events_by_type.py (hunk 7 above)
Comment (andrei-neon) [reply integration/tests/basic/indexer/test_events_by_type.py]: Done ✅

File: integration/tests/basic/indexer/test_events.py (hunk 10 above)
Comment (andrei-neon) [reply integration/tests/basic/indexer/test_events.py:193]: Fixed ✅

File: integration/tests/basic/helpers/rpc_checks.py (hunk 8 above)
Comment (andrei-neon) [reply integration/tests/basic/helpers/rpc_checks.py]: I refactored this function a little bit (match is remaining but events are checked in a loop).

File: integration/tests/basic/indexer/conftest.py (hunk 9 above)
Comment (andrei-neon) [reply integration/tests/basic/indexer/conftest.py]: Fixed ✅

File: integration/tests/basic/helpers/rpc_checks.py [hunk 11]
```diff
@@ -201,4 +206,172 @@ def assert_equal_fields(result, comparable_object, comparable_fields, keys_mappi
...
+def count_events(
+    neon_trx_receipt: NeonGetTransactionResult,
+    ignored_events=[NeonEventType.InvalidRevision, NeonEventType.StepReset],
+    is_removed=False,
+):
+    events = neon_trx_receipt.get_all_events(ignore_events=ignored_events, is_removed=is_removed)
+    return Counter([event.neonEventType for event in events])
+
+
+def assert_event_by_type(neon_trx_receipt: NeonGetTransactionResult):
```
Comment (kristinaNikolaevaa) [inline integration/tests/basic/helpers/rpc_checks.py]: assert_eventS_by_type?

File: integration/tests/basic/indexer/test_events.py [hunk 12]
```diff
@@ -0,0 +1,238 @@
...
+        resp = self.web3_client.send_transaction(sender_account, tx)
+
+        response = json_rpc_client.get_neon_trx_receipt(resp["transactionHash"])
+        validated_response = NeonGetTransactionResult(**response)
+
+        assert count_events(validated_response) == Counter({"EnterCall": 1, "ExitStop": 1, "Return": 1})
+        assert_events_order(validated_response)
+        assert_event_by_type(validated_response)
+
+    def test_basic_transtaction_receipt_transaction_hash(self, json_rpc_client):
```
Comment (kristinaNikolaevaa) [inline integration/tests/basic/indexer/test_events.py]: test_field_values_for_trx_with_transfer?

File: integration/tests/basic/indexer/test_events.py [hunk 13]
```diff
@@ -0,0 +1,238 @@
...
+
+@allure.feature("JSON-RPC validation")
+@allure.story("Verify events")
+@pytest.mark.usefixtures("accounts", "web3_client")
+@pytest.mark.neon_only
+class TestEvents:
+    web3_client: NeonChainWeb3Client
+    accounts: EthAccounts
+
+    def test_basic_transtaction_receipt(self, json_rpc_client):
```
Comment (kristinaNikolaevaa) [inline integration/tests/basic/indexer/test_events.py]: test_events_for_trx_with_transfer?

File: integration/tests/basic/indexer/test_events.py [hunk 14]
```diff
@@ -0,0 +1,238 @@
...
+        assert_event_field(validated_response, NeonEventType.Return, "transactionHash", transaction_hash)
+
+        assert_event_field(validated_response, NeonEventType.EnterCall, "address", receiver_account.address)
+        assert_event_field(validated_response, NeonEventType.ExitStop, "address", receiver_account.address)
+        assert_event_field(validated_response, NeonEventType.Return, "address", None)
+
+        assert count_events(validated_response) == Counter({"EnterCall": 1, "ExitStop": 1, "Return": 1})
+        assert_event_by_type(validated_response)
+
+    def test_basic_transtaction_receipt_contract(self, json_rpc_client, event_caller_contract):
```
Comment (kristinaNikolaevaa) [inline integration/tests/basic/indexer/test_events.py]: test_events_for_trx_with_logs

File: integration/tests/basic/indexer/test_events.py [hunk 15]
```diff
@@ -0,0 +1,238 @@
...
+        assert count_events(validated_response) == Counter({"EnterCall": 1, "ExitStop": 1, "Return": 1, "Log": 1})
+
+        assert_event_field(validated_response, NeonEventType.EnterCall, "address", validated_response.result.to)
+        assert_event_field(validated_response, NeonEventType.ExitStop, "address", validated_response.result.to)
+        assert_event_field(validated_response, NeonEventType.Log, "address", validated_response.result.to)
+        assert_event_field(validated_response, NeonEventType.Return, "address", None)
+
+        assert_event_by_type(validated_response)
+
+    def test_basic_transtaction_receipt_nested_call(self, json_rpc_client, nested_call_contracts):
```
Comment (kristinaNikolaevaa) [inline integration/tests/basic/indexer/test_events.py]: what is BASIC here

File: integration/tests/basic/indexer/test_events.py [hunk 16]
```diff
@@ -0,0 +1,238 @@
...
+        resp = self.web3_client.send_transaction(sender_account, instruction_tx)
+
+        response = json_rpc_client.get_neon_trx_receipt(resp["transactionHash"])
+        validated_response = NeonGetTransactionResult(**response)
+
+        assert count_events(validated_response) == Counter({"EnterCall": 4, "ExitStop": 2, "Return": 1, "Log": 2})
+        assert_events_order(validated_response)
+        assert_event_by_type(validated_response)
+
+    def test_contract_interative_operation(self, counter_contract, account_with_all_tokens, json_rpc_client):
```
Comment (kristinaNikolaevaa) [inline integration/tests/basic/indexer/test_events.py]: iterative transaction

File: integration/tests/basic/indexer/test_events.py [hunk 17]
```diff
@@ -0,0 +1,238 @@
...
+        tx = self.web3_client.make_raw_tx(account_with_all_tokens.address)
```
Comment (kristinaNikolaevaa) [inline integration/tests/basic/indexer/test_events.py]: please use simple account from accounts ficture
account_with_all_tokens needed for testing different chains
It takes some. time to create it and here is no needs for it

File: integration/tests/basic/indexer/test_events.py [hunk 18]
```diff
@@ -0,0 +1,238 @@
...
+
+        instruction_tx = counter_contract.functions.moreInstruction(0, 1000).build_transaction(tx)
+        resp = self.web3_client.send_transaction(account_with_all_tokens, instruction_tx)
+        response = json_rpc_client.get_neon_trx_receipt(resp["transactionHash"])
+        validated_response = NeonGetTransactionResult(**response)
+
+        assert len(validated_response.result.solanaTransactions) > 1
+        assert count_events(validated_response) == Counter({"EnterCall": 1, "ExitStop": 1, "Return": 1})
```
Comment (kristinaNikolaevaa) [inline integration/tests/basic/indexer/test_events.py]: Iterative transaction has only 3 events??

File: integration/tests/conftest.py [hunk 19]
```diff
@@ -508,3 +495,59 @@ def neon_price(web3_client_session) -> float:
     price = web3_client_session.get_token_usd_gas_price()
     with allure.step(f"NEON price {price}$"):
         return price
+
+
+@pytest.fixture(scope="class")
+def events_checker_contract(web3_client, accounts) -> tp.Any:
+    contract, _ = web3_client.deploy_and_get_contract("common/tracer/ContractCaller", "0.8.15", account=accounts[0])
```
Comment (kristinaNikolaevaa) [inline integration/tests/conftest.py]: let's move it from tracer folder

File: integration/tests/conftest.py [hunk 20]
```diff
@@ -508,3 +495,59 @@ def neon_price(web3_client_session) -> float:
...
+        "0.8.10",
+        sender_account,
+        contract_name="DeployRecursionFactory",
+        constructor_args=[3],
+    )
+    yield contract
+
+
+@pytest.fixture(scope="function")
+def destroyable_contract(accounts, web3_client):
```
Comment (kristinaNikolaevaa) [inline integration/tests/conftest.py:540]: please delete it from TestSelfDestructOpcode class than you moved it

File: integration/tests/economy/conftest.py [hunk 21]
```diff
@@ -19,14 +18,6 @@ def sol_client_tx_v2(pytestconfig: Config):
     return client
 
 
-@pytest.fixture(scope="class")
-def counter_contract(account_with_all_tokens, client_and_price, web3_client_sol, web3_client):
```
Comment (kristinaNikolaevaa) [inline integration/tests/economy/conftest.py]: for operator tests it is important to have counter_contract as it was (with client_and_price)

File: integration/tests/conftest.py (hunk 20 above)
Comment (kristinaNikolaevaa) [reply integration/tests/conftest.py:540]: the same question about all new fixtures

File: integration/tests/basic/helpers/rpc_checks.py (hunk 11 above)
Comment (andrei-neon) [reply integration/tests/basic/helpers/rpc_checks.py]: Fixed ✅

File: integration/tests/basic/indexer/test_events.py (hunk 12 above)
Comment (andrei-neon) [reply integration/tests/basic/indexer/test_events.py]: Fixed ✅

File: integration/tests/basic/indexer/test_events.py (hunk 13 above)
Comment (andrei-neon) [reply integration/tests/basic/indexer/test_events.py]: Fixed ✅

File: integration/tests/basic/indexer/test_events.py (hunk 14 above)
Comment (andrei-neon) [reply integration/tests/basic/indexer/test_events.py]: Fixed ✅

File: integration/tests/basic/indexer/test_events.py (hunk 17 above)
Comment (andrei-neon) [reply integration/tests/basic/indexer/test_events.py]: Fixed ✅

File: integration/tests/basic/indexer/test_events.py (hunk 18 above)
Comment (andrei-neon) [reply integration/tests/basic/indexer/test_events.py]: Added Log events ✅

File: integration/tests/conftest.py (hunk 19 above)
Comment (andrei-neon) [reply integration/tests/conftest.py]: Fixed ✅

File: integration/tests/conftest.py (hunk 20 above)
Comment (andrei-neon) [reply integration/tests/conftest.py:540]: Fixed ✅

File: integration/tests/basic/indexer/test_events.py (hunk 16 above)
Comment (andrei-neon) [reply integration/tests/basic/indexer/test_events.py]: Fixed ✅

File: integration/tests/economy/conftest.py (hunk 21 above)
Comment (andrei-neon) [reply integration/tests/economy/conftest.py]: Reverted ✅

File: integration/tests/basic/indexer/test_events.py (hunk 15 above)
Comment (andrei-neon) [reply integration/tests/basic/indexer/test_events.py]: Renamed ✅

**Review by kristinaNikolaevaa — APPROVED** [review]

**Pull request merged**